File: MyCamera/plugins/cordova-plugin-mfp-push/src/ios/CDVMFPPush.m
/*
   Licensed Materials - Property of IBM

   (C) Copyright 2016 IBM Corp.

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */
 
#import "CDVMFPPush.h"


// private methods declaration
@interface CDVMFPPush()
    void _setValueIfExists(NSString *key, NSDictionary *source, NSMutableDictionary *dest);
    NSSet* _createCategories(NSArray *categoryArray);
    NSArray* _createActionsDictionary(NSArray *actionArray);
    NSArray* _filterActions(NSArray *actionIds,NSArray* actions);
@end

@implementation CDVMFPPush

typedef enum MFPUIUserNotificationActionContext : NSUInteger {
    MFPUIUserNotificationActionContextDefault,
    MFPUIUserNotificationActionContextMinimal
} MFPUIUserNotificationActionContext;

typedef enum MFPUIUserNotificationActivationMode : NSUInteger {
    MFPUIUserNotificationActivationModeForeground,
    MFPUIUserNotificationActivationModeBackground
} MFPUIUserNotificationActivationMode;


@synthesize registerCallbackId;
@synthesize registerCommandDelegate;
@synthesize notifCallbackId;
@synthesize notifCommandDelegate;


- (void) pluginInitialize {
    [super pluginInitialize];

    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didRegisterForRemoteNotifications:) name:CDVMFPPushDidRegisterForRemoteNotificationsWithDeviceToken object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didFailToRegisterForRemoteNotifications:) name:CDVMFPPushDidFailToRegisterForRemoteNotificationsWithError object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didReceiveRemoteNotification:) name:CDVMFPPushDidReceiveRemoteNotification object:nil];
    
    [self processRemoteNotificationsOnPluginInitialization];
}

- (void)processRemoteNotificationsOnPluginInitialization
{

    NSMutableDictionary *launchOptions = ((MFPAppDelegate *)[self appDelegate]).launchOptions;
    @synchronized(launchOptions) {
        NSDictionary *userInfo = [launchOptions objectForKey:UIApplicationLaunchOptionsRemoteNotificationKey];
        if (userInfo) {
            [self processRemoteNotification:userInfo];
            // Remove remote notifications info from the launch options to eliminate duplicate handling
            [launchOptions removeObjectForKey:UIApplicationLaunchOptionsRemoteNotificationKey];
        }
    }
}


/*
 * Initializes MFPPush instance
 */

- (void) initialize: (CDVInvokedUrlCommand*)command {
    
    CDVPluginResult *result = nil;
    
    NSArray* arguments = [[NSArray alloc]initWithArray:command.arguments];
    id timeoutArg = [arguments objectAtIndex:0];
    NSTimeInterval timeout = [timeoutArg doubleValue];
    MFPPush *push = [MFPPush sharedInstance];
    if(push != nil) {
        [push initialize:timeout];
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];

}

/*
 * Checks whether push notification is supported.
 */
- (void) isPushSupported: (CDVInvokedUrlCommand*)command {
    
    NSString *isSupported = @"true";
    BOOL isPushSupported = [[MFPPush sharedInstance] isPushSupported];
    if(!isPushSupported) {
        isSupported = @"false";
    }
    
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:isSupported];
    
    [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
}


/*
 * Registers the device with APNs
 */
-(void) registerDevice: (CDVInvokedUrlCommand*)command {
    
    registerCallbackId = command.callbackId;
    registerCommandDelegate = self.commandDelegate;
    NSDictionary* options = NULL;
    NSDictionary* iosOptions = NULL;

    
    NSArray* arguments = [[NSArray alloc]initWithArray:command.arguments];
    NSLog(@"Value of arguments count is %lu", (unsigned long)arguments.count);
    if ([arguments count] > 0) {
        NSLog(@"arguments are not null");
        options = [[NSDictionary alloc]initWithDictionary:[arguments objectAtIndex:0]];
        
        iosOptions = _setDeviceOptions(options);
    }
    
    [self.commandDelegate runInBackground:^{
        
        [[MFPPush sharedInstance]registerDevice:iosOptions completionHandler:^(WLResponse *response, NSError *error){
            CDVPluginResult *result = nil;
            if (error != nil) {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
            } else {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:response.responseText];
            }
            [registerCommandDelegate sendPluginResult:result callbackId:registerCallbackId];
            
        }];
        
        
    }];

}


/*
 * Unregisters the device with MFP Push Notification Server
 */
-(void) unregisterDevice: (CDVInvokedUrlCommand*)command  {
    
    [self.commandDelegate runInBackground:^{
        [[MFPPush sharedInstance] unregisterDevice:^(WLResponse *response, NSError *error) {
            CDVPluginResult *result = nil;
            if (error != nil) {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
            } else {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:response.description];
            }
            [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
        }];
        
    }];
}

/*
 * Subscribes to a particular backend mobile application Tag(s)
 */
-(void) subscribe: (CDVInvokedUrlCommand*)command  {
    
    [self.commandDelegate runInBackground:^{
        id tags = command.arguments[0];
        if([tags isKindOfClass:[NSString class]]){
            NSArray *tagsArray = [tags componentsSeparatedByString:@","];
            [[MFPPush sharedInstance] subscribe:tagsArray completionHandler:^(WLResponse *response, NSError *error) {
                CDVPluginResult *result = nil;
                if (error != nil) {
                    result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
                } else {
                    result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:response.responseText];
                }
                [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
            }];
        } else {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Tag Parameter is Invalid."];
            [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
        }
    }];
}

/*
 * Unsubscribes from particular backend mobile application Tag(s)
 */
-(void) unsubscribe: (CDVInvokedUrlCommand*)command  {
    
    [self.commandDelegate runInBackground:^{
        id tags = command.arguments[0];
        if([tags isKindOfClass:[NSString class]]){
            NSArray *tagsArray = [tags componentsSeparatedByString:@","];
            [[MFPPush sharedInstance] unsubscribe:tagsArray completionHandler:^(WLResponse *response, NSError *error) {
                CDVPluginResult *result = nil;
                if (error != nil) {
                    result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
                } else {
                    result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:response.responseText];
                }
                [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
            }];
        } else {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Tag Parameter is Invalid."];
            [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
        }
    }];
}
/*
 * Gets all the available Tags for the backend mobile application
 */
-(void) getTags: (CDVInvokedUrlCommand*)command {
    
    [self.commandDelegate runInBackground:^{
        
        [[MFPPush sharedInstance] getTags:^(WLResponse *response, NSError *error) {
            CDVPluginResult *result = nil;
            if (error != nil) {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
            } else {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:response.availableTags];
            }
            [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
            
        }];
        
    }];
}


/*
 * Gets the Tags that are subscribed by the device
 */
-(void) getSubscriptions: (CDVInvokedUrlCommand*)command {
    
    [self.commandDelegate runInBackground:^{
        
        [[MFPPush sharedInstance] getSubscriptions:^(WLResponse *response, NSError *error) {
            CDVPluginResult *result = nil;
            if (error != nil) {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
            } else {
                result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:response.subscriptions];
            }
            [self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];
            
        }];
        
    }];
}

-(void) registerNotificationsCallback: (CDVInvokedUrlCommand*)command  {
    
    notifCallbackId = command.callbackId;
    notifCommandDelegate = self.commandDelegate;
    
    [self dispatch];
}

-(void) backgroundJobDone:(CDVInvokedUrlCommand *)command {
    NSString *id = [[NSString alloc] initWithString:[command.arguments objectAtIndex:0]];
    int jsResult = [[command.arguments objectAtIndex:1] intValue];
    MFPAppDelegate* delegate = (MFPAppDelegate *)[self appDelegate];
    CompletionHandler handler = (CompletionHandler)[delegate.completionHandlers objectForKey:id];
    if(handler){
        UIBackgroundFetchResult result = UIBackgroundFetchResultNoData;
        if(jsResult < 0){
            result = UIBackgroundFetchResultFailed;
        }
        if(jsResult > 0){
            result = UIBackgroundFetchResultNewData;
        }
        handler(result);
        [delegate.completionHandlers removeObjectForKey:id];
    }
}

// Internal functions

/*
 * Function called after registered for remote notifications. Registers device token from APNs with MFP Push Server.
 */
-(void) didRegisterForRemoteNotifications: (NSNotification*) notification {
    
    NSData* deviceToken = [notification object];
    
    if (registerCallbackId == nil) {
        return;
    }
    
    
    [registerCommandDelegate runInBackground:^{
        
        [[MFPPush sharedInstance] sendDeviceToken:deviceToken];
        
    }];
}

-(void) didFailToRegisterForRemoteNotifications: (NSNotification*) notification {
    

    NSError* error = [notification object];
    
    if (registerCallbackId == nil) {
        return;
    }
    
    [registerCommandDelegate runInBackground:^{
        
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
        [registerCommandDelegate sendPluginResult:result callbackId:registerCallbackId];
    }];
}

-(void) didReceiveRemoteNotification: (NSNotification*) notification {
    NSDictionary* userInfo = [notification object];
    [self processRemoteNotification:userInfo];
    
}


-(void) processRemoteNotification: (NSDictionary*) notification {
    
    NSMutableDictionary *notif = [[NSMutableDictionary alloc] init];
    NSDictionary *aps = [notification objectForKey:@"aps"];
    _setValueIfExists(@"alert", aps, notif);
    _setValueIfExists(@"badge", aps, notif);
    _setValueIfExists(@"sound", aps, notif);
    _setValueIfExists(@"content-available",aps,notif);
    //If the props contains content-available Initialize to -1 and overwrite with value if callback-id is available
    if([notif objectForKey:@"content-available"]){
        [notif setValue:@"-1" forKey:@"callback-id"];
    }
    _setValueIfExists(@"callback-id",aps,notif);
    _setValueIfExists(@"category",aps,notif);
    _setValueIfExists(@"action-id",aps,notif);
    NSDictionary *payload = [notification objectForKey:@"payload"];
    [notif setValue:payload forKey:@"payload"];
    
    [self queueMessage:notif];
}

- (void) sendToJSCallback: (NSMutableDictionary *) notification  {

    [notifCommandDelegate runInBackground:^{
        CDVPluginResult *result = nil;
        if (notification == nil) {
            result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in receiving notification"];
            [result setKeepCallbackAsBool:true];
        }
        else {
            result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:notification];
            [result setKeepCallbackAsBool:true];
            
        }
        [notifCommandDelegate sendPluginResult:result callbackId:notifCallbackId];
        
    }];
    
}

- (void) queueMessage:(NSMutableDictionary *)notification {
    @synchronized (pending){
        if (!pending) {
            pending = [[NSMutableArray alloc] initWithCapacity:8];
        }
        [pending addObject: notification];
        NSString *isSilent = @"false";
        NSNumber *contentAvialable = [notification objectForKey:@"content-available"];
        if([contentAvialable intValue] == 1){
            if(![notification objectForKey:@"alert"]){
                isSilent = @"true";
            }
            else{
                UIApplicationState state = [[UIApplication sharedApplication] applicationState];
                if(state == UIApplicationStateBackground){
                    isSilent = @"true";
                }
            }
        }
        // Workaround for javascript functions that do some UI.
        // We need to delay a bit to let view appear, otherwise the application will become forzen.
        //TODO: validate when webview is available and not use delay
        if (notifCallbackId != nil) {
            [self performSelector:@selector(dispatchSelector:) withObject:isSilent afterDelay:1];
        }
    }
}

- (void) dispatchSelector: (NSString *) isSilent {
    @synchronized (pending) {
        [self dispatch];
        if([isSilent isEqualToString:@"false"]){
            // Hide notifications from notification center
            NSInteger badgeNum = [[UIApplication sharedApplication] applicationIconBadgeNumber];
            [[UIApplication sharedApplication] setApplicationIconBadgeNumber:1];
            [[UIApplication sharedApplication] setApplicationIconBadgeNumber:0];
            [[UIApplication sharedApplication] setApplicationIconBadgeNumber:badgeNum];
        }
    }
}

- (void) dispatch {
    @synchronized (pending) {
        if (pending)	{
            for (NSInteger i = 0, n = pending.count; i < n; ++i) {
                NSMutableDictionary *notification = [pending objectAtIndex:i];
                
                BOOL setApplicationBadgeNo = NO;
                UIApplication *application = [UIApplication sharedApplication];
                if([application respondsToSelector:@selector(registerUserNotificationSettings:)]){
                    id userNotificationSettings = ((id (*) (id, SEL)) objc_msgSend) (application,@selector(currentUserNotificationSettings));
                    NSUInteger types = ((NSUInteger (*) (id, SEL)) objc_msgSend)(userNotificationSettings, @selector(types));
                    if(types & (1 << 0)){
                        setApplicationBadgeNo = YES;
                    }
                }
                else{
                    if ([application enabledRemoteNotificationTypes] & UIRemoteNotificationTypeBadge) {
                        setApplicationBadgeNo = YES;
                    }
                }
                
                if (setApplicationBadgeNo) {
                    application.applicationIconBadgeNumber = [((NSString *)[notification valueForKey:@"badge"]) intValue];
                }
                [self sendToJSCallback: notification];
            }
            
            [pending removeAllObjects];
        }
    }
}

void _setValueIfExists(NSString *key, NSDictionary *source, NSMutableDictionary *dest) {
    if (!source) return;
    id obj = [source valueForKey:key];
    if (obj) [dest setValue:obj forKey:key];
}

NSSet* _createCategories(NSArray *categoryArray){
    
    NSMutableSet* categories = nil;
    if([categoryArray count] > 0) {
        categories = [[NSMutableSet alloc] init];
        for(id categoryObj in categoryArray){
            NSDictionary *categoryDic = [[NSDictionary alloc] initWithDictionary:categoryObj];
            
            id category = [[NSClassFromString(@"UIMutableUserNotificationCategory") alloc] init];
            
            [category setValue:(NSString *)[categoryDic valueForKey:@"id"] forKey:@"identifier"];
            
            //Create actions array object
            NSArray *actionArray = [[NSArray alloc] initWithArray:[categoryDic valueForKey:@"actions"]];
            NSArray *actions = _createActionsDictionary(actionArray);
            
            
            //Set the default context
            NSArray *defaultContextActionsJSON = [[NSArray alloc] initWithArray:[categoryDic valueForKey:@"defaultContextActions"]];
            NSArray *defaultContextActions = _filterActions(defaultContextActionsJSON, actions);
            if([defaultContextActions count] > 0){
                ((void (*) (id, SEL, id, NSUInteger))objc_msgSend)(category, @selector(setActions:forContext:),defaultContextActions,MFPUIUserNotificationActionContextDefault);
            }
            else{
                ((void (*) (id, SEL, id, NSUInteger))objc_msgSend)(category, @selector(setActions:forContext:),actions ,MFPUIUserNotificationActionContextDefault);
            }
            
            
            //Set the minimal context
            NSArray *minimalContextActionsJSON = [[NSArray alloc] initWithArray:[categoryDic valueForKey:@"minimalContextActions"]];
            NSArray *minimalContextActions = _filterActions(minimalContextActionsJSON, actions);
            if([minimalContextActionsJSON count] > 0){
                ((void (*) (id, SEL, id, NSUInteger))objc_msgSend)(category, @selector(setActions:forContext:),minimalContextActions,MFPUIUserNotificationActionContextMinimal);
            }
            else{
                ((void (*) (id, SEL, id, NSUInteger))objc_msgSend)(category, @selector(setActions:forContext:),actions,MFPUIUserNotificationActionContextMinimal);
            }
            
            [categories addObject:category];
        }
    }
    return categories;
}

NSArray* _createActionsDictionary(NSArray *actionArray){
    
    NSMutableArray *actions = [[NSMutableArray alloc] init];
    for(id actionObj in actionArray){
        NSDictionary *actionDic = [[NSDictionary alloc] initWithDictionary:actionObj];
        
        id action = [[NSClassFromString(@"UIMutableUserNotificationAction") alloc] init];
        
        [action setValue:(NSString *)[actionDic valueForKey:@"id"] forKey:@"identifier"];
        [action setValue: (NSString *)[actionDic valueForKey:@"title"] forKey:@"title"];
        [action setValue:[NSNumber numberWithInt:MFPUIUserNotificationActivationModeForeground] forKey:@"activationMode"];
        
        if([actionDic objectForKey:@"mode"] != nil){
            NSInteger mode = [[actionDic valueForKey:@"mode"] integerValue];
            if(mode  == 0){
                [action setValue:[NSNumber numberWithInt:MFPUIUserNotificationActivationModeBackground] forKey:@"activationMode"];
            }
        }
        
        if([actionDic objectForKey:@"destructive"] != nil){
            BOOL destructive = [[actionDic valueForKey:@"destructive"] boolValue];
            [action setValue:[NSNumber numberWithBool:destructive] forKey:@"destructive"];
        }
        
        if([actionDic objectForKey:@"authenticationRequired"] != nil){
            BOOL requireAuthentication = [[actionDic valueForKey:@"authenticationRequired"] boolValue];
            [action setValue:[NSNumber numberWithBool:requireAuthentication] forKey:@"authenticationRequired"];
        }
        
        [actions addObject:action];
    }
    return actions;
}

NSArray* _filterActions(NSArray *actionIds,NSArray* actions){
    NSMutableArray *filteredActions = [[NSMutableArray alloc]init];
    if([actionIds count] > 0){
        for(NSString *identifier in actionIds){
            for(id action in actions){
                NSString *actionId = (NSString *) [action valueForKey:@"identifier"];
                if([identifier isEqualToString:actionId]){
                    [filteredActions addObject:action];
                    break;
                }
            }
            
        }
    }
    return filteredActions;
}

/*
 * Sets the push notification options.
 */
NSDictionary* _setDeviceOptions(NSDictionary *options) {
    
    CDVPluginResult *result = nil;
    NSDictionary* iosOptions = nil;
    NSObject *phoneNumber = nil;
    if(options != nil && ![options isKindOfClass:[NSNull class]]) {
        
        iosOptions = [options objectForKey:@"ios"];
        phoneNumber = [options objectForKey:@"phoneNumber"];
        
        if(phoneNumber != nil) {
            if(iosOptions == nil) {
                iosOptions = [[NSMutableDictionary alloc] init];
            }
            [iosOptions setValue:phoneNumber forKey:@"phoneNumber"];
        }
        
        if(iosOptions != nil && ![iosOptions isKindOfClass:[NSNull class]]) {

            NSSet *categories = _createCategories([iosOptions objectForKey:@"categories"]);
            [iosOptions setValue:categories forKey:@"categories"];
            
            result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        }
        
    } else {
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    
    /*[self.commandDelegate sendPluginResult:result callbackId:[command callbackId]];*/
    return iosOptions;
}




@end

File: MyCamera/plugins/cordova-plugin-mfp-push/src/android/CDVMFPPush.java
/*
   Licensed Materials - Property of IBM

   (C) Copyright 2016 IBM Corp.

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

package com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.push;

import com.worklight.common.Logger;
import com.ibm.mobilefirstplatform.clientsdk.android.push.api.MFPPush;
import com.ibm.mobilefirstplatform.clientsdk.android.push.api.MFPSimplePushNotification;
import com.ibm.mobilefirstplatform.clientsdk.android.push.api.MFPPushException;
import com.ibm.mobilefirstplatform.clientsdk.android.push.api.MFPPushNotificationListener;
import com.ibm.mobilefirstplatform.clientsdk.android.push.api.MFPPushResponseListener;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.Arrays;

public class CDVMFPPush extends CordovaPlugin {

    private static final Logger pushLogger = Logger.getInstance("com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.push.CDVMFPPush");

    private static CallbackContext notificationCallback;

    private static MFPPushNotificationListener notificationListener;

    private static boolean ignoreIncomingNotifications = false;

    @Override
    public boolean execute(String action, JSONArray args, final CallbackContext callbackContext) throws JSONException {
        pushLogger.debug("execute() : action = " + action);

        if ("initialize".equals(action)) {
        	int timeout = args.getInt(0);
            this.initialize(timeout, callbackContext);

            return true;
        } else if ("isPushSupported".equals(action)) {
            this.isPushSupported(callbackContext);

            return true;
        }  else if ("registerDevice".equals(action)) {
            JSONObject options = null;
            if(args.length() > 0)
                options = (JSONObject)args.get(0);

            this.registerDevice(callbackContext,options);
            return true;
        } else if ("unregisterDevice".equals(action)) {
            this.unregisterDevice(callbackContext);

            return true;
        } else if ("getSubscriptions".equals(action)) {
            this.getSubscriptions(callbackContext);

            return true;
        } else if ("getTags".equals(action)) {
            this.getTags(callbackContext);

            return true;
        } else if ("subscribe".equals(action)) {
            String[] tags = args.getString(0).split(",");
            this.subscribe(tags, callbackContext);

            return true;
        } else if ("unsubscribe".equals(action)) {
			String[] tags = args.getString(0).split(",");
            this.unsubscribe(tags, callbackContext);

            return true;
        } else if ("registerNotificationsCallback".equals(action)) {
            this.registerNotificationsCallback(callbackContext);

            return true;
        }
        return false;
    }
    
    /**
     * Initializes MFPPush instance
     * @param timeout request timeout in seconds
     * @param callbackContext Javascript callback
     */    
    private void initialize(int timeout, final CallbackContext callbackContext) {
    	try  {
    		MFPPush.getInstance().initialize(this.cordova.getActivity().getApplicationContext(), timeout);
    		callbackContext.success();
    	} catch (Exception e) {
    		callbackContext.error(e.toString());
    	}
    }
    
    
    /**
     * Checks whether push notification is supported.
     * @param callbackContext Javascript callback
     */    
    private void isPushSupported(final CallbackContext callbackContext) {
    	try  {
        	boolean isPushSupported = MFPPush.getInstance().isPushSupported();
        	callbackContext.success(Boolean.toString(isPushSupported));
    	} catch (Exception e) {
    		callbackContext.error(e.toString());
    	}
    }    
    /**
     * Sets the push notification options.
     * 	
     * @param options:
	 *        ios: { alert: boolean, badge: boolean, sound: boolean, categories: Object[] }
	 *        android: {}
     * @param callbackContext Javascript callback
     */ 
    private void setOptions(final String options, final CallbackContext callbackContext) {
    	pushLogger.debug("In setOptions");
    	callbackContext.success();
    }
    
    /**
     * Registers the device with the push service
     * @param callbackContext Javascript callback
     * @param options:
	 *        ios: { alert: boolean, badge: boolean, sound: boolean, categories: Object[] }
	 *        android: {}
     * 		  phoneNumber: String	 
     * 
     */
    private void registerDevice(final CallbackContext callbackContext, final JSONObject options) {
    	pushLogger.debug("In registerDevice");
        JSONObject optionsJSON = null;
        String phoneNumber = null;
        try {

            if(options != null) {
                if(options.has("android"))
                    optionsJSON = (JSONObject)options.get("android");
                if(options.has("phoneNumber"))
                    phoneNumber = (String) options.get("phoneNumber");
                if(phoneNumber != null && optionsJSON == null) {
                    optionsJSON = new JSONObject();
                }
                if(phoneNumber != null) {
                    optionsJSON.put("phoneNumber", options.get("phoneNumber"));
                }
            }
        }
        catch(JSONException ex)
        {
            pushLogger.debug("In registerDevice json parsing exception");
        }
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MFPPush.getInstance().registerDevice(options, new MFPPushResponseListener<String>() {
                    @Override
                    public void onSuccess(String s) {
                        pushLogger.debug("registerDevice() Success : " + s);
                        callbackContext.success(s);
                    }
                    @Override
                    public void onFailure(MFPPushException ex) {
                        pushLogger.debug("registerDevice() Error : " + ex.toString());
                        callbackContext.error(ex.toString());
                    }
                });
            }
        });

    }

    /**
     * Unregister the device from push service
     * @param callbackContext Javascript callback
     */
    private void unregisterDevice(final CallbackContext callbackContext) {
        pushLogger.debug("In unregisterDevice");

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MFPPush.getInstance().unregisterDevice(new MFPPushResponseListener<String>() {
                    @Override
                    public void onSuccess(String s) {
                        pushLogger.debug("unregisterDevice() Success : " + s);
                        callbackContext.success(s);
                    }
                    @Override
                    public void onFailure(MFPPushException ex) {
                        pushLogger.debug("unregisterDevice() Error : " + ex.toString());
                        callbackContext.error(ex.toString());
                    }
                });
            }
        });

    }

    /**
     * Get the list of available tags that the device can subscribe to
     * @param callbackContext Javascript callback
     */
    private void getTags(final CallbackContext callbackContext) {
        pushLogger.debug("In getTags");

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MFPPush.getInstance().getTags(new MFPPushResponseListener<List<String>>() {
                    @Override
                    public void onSuccess(List<String> tags) {
                        pushLogger.debug("getTags() Success : " + tags);
                        callbackContext.success(new JSONArray(tags));
                    }

                    @Override
                    public void onFailure(MFPPushException ex) {
                        pushLogger.debug("getTags() Error : " + ex.toString());
                        callbackContext.error(ex.toString());
                    }
                });
            }
        });

    }

    /**
     * Get the list of tags subscribed to
     * @param callbackContext Javascript callback
     */
    private void getSubscriptions(final CallbackContext callbackContext) {
        pushLogger.debug("In getSubscriptions");

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                MFPPush.getInstance().getSubscriptions(new MFPPushResponseListener<List<String>>() {
                    @Override
                    public void onSuccess(List<String> tags) {
                        pushLogger.debug("getSubscriptions() Success : " + tags);
                        callbackContext.success(new JSONArray(tags));
                    }

                    @Override
                    public void onFailure(MFPPushException ex) {
                        pushLogger.debug("getSubscriptions() Error : " + ex.toString());
                        callbackContext.error(ex.toString());
                    }
                });
            }
        });

    }

    /**
     * Subscribes to the given tag(s)
     * @param tags
     * @param callbackContext Javascript callback
     */
    private void subscribe(final String[] tags, final CallbackContext callbackContext) {
        pushLogger.debug("In subscribe");

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                    MFPPush.getInstance().subscribe(tags, new MFPPushResponseListener<String[]>() {
                        @Override
                        public void onSuccess(String[] s) {
                            pushLogger.debug("subscribe() Success : " + s);
							List<String> tagNames = Arrays.asList(s);
                            callbackContext.success(new JSONArray(tagNames));
                        }

                        @Override
                        public void onFailure(MFPPushException ex) {
                            pushLogger.debug("subscribe() Error : " + ex.toString());
                            callbackContext.error(ex.toString());
                        }
                     });
            }
        });

    }

    /**
     * Unsubscribes to the given tag(s)
     * @param tags
     * @param callbackContext Javascript callback
     */
    private void unsubscribe(final String[] tags, final CallbackContext callbackContext) {
        pushLogger.debug("In unsubscribe");

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                    MFPPush.getInstance().unsubscribe(tags, new MFPPushResponseListener<String[]>() {
                        @Override
                        public void onSuccess(String[] s) {
                            pushLogger.debug("unsubscribe() Success : " + s);
							List<String> tagNames = Arrays.asList(s);
                            callbackContext.success(new JSONArray(tagNames));
                        }

                        @Override
                        public void onFailure(MFPPushException ex) {
                            pushLogger.debug("unsubscribe() Error : " + ex.toString());
                            callbackContext.error(ex.toString());
                        }
                     });
            }
        });

    }

    private void registerNotificationsCallback(final CallbackContext callbackContext) {
        pushLogger.debug("In registerNotificationsCallback");

        notificationCallback = callbackContext;

        if(!ignoreIncomingNotifications) {

            cordova.getThreadPool().execute(new Runnable() {
                public void run() {
                    notificationListener = new MFPPushNotificationListener() {
                        @Override
                        public void onReceive(final MFPSimplePushNotification message) {
                            try {
                                pushLogger.debug("Push notification received: " + message.toString());

                                JSONObject notification = new JSONObject();

                                notification.put("alert", message.getAlert());
                                notification.put("payload", message.getPayload());

                                PluginResult result = new PluginResult(PluginResult.Status.OK, notification);
                                result.setKeepCallback(true);
                                notificationCallback.sendPluginResult(result);
                            } catch(JSONException e) {
                                PluginResult result = new PluginResult(PluginResult.Status.ERROR, e.toString());
                                result.setKeepCallback(true);
                                notificationCallback.sendPluginResult(result);
                            }
                        }
                    };

                    MFPPush.getInstance().listen(notificationListener);
                }
            });

        } else {
            pushLogger.warn("Notification handling is currently off. Turn it back on by calling setIgnoreIncomingNotifications(true)");
            callbackContext.error("Error: Called registerNotificationsCallback() after IgnoreIncomingNotifications was set");
        }
    }

    /**
     * Change the plugin's default behavior to ignore handling push notifications
     * @param ignore boolean parameter for the 'ignore notifications' behavior
     */
    public static void setIgnoreIncomingNotifications(boolean ignore) {
        pushLogger.debug("In setIgnoreIncomingNotifications : ignore = " + ignore);
        ignoreIncomingNotifications = ignore;

        if(notificationListener != null) {
            if(ignore) {
                MFPPush.getInstance().hold();
            } else {
                MFPPush.getInstance().listen(notificationListener);
            }
        }

    }

    @Override
    public void onResume(boolean multitasking) {
        super.onResume(multitasking);
        pushLogger.debug("In onResume");

        if (!ignoreIncomingNotifications && MFPPush.getInstance() != null) {
            MFPPush.getInstance().listen(notificationListener);
        }
    }

    @Override
    public void onPause(boolean multitasking) {
        super.onPause(multitasking);
        pushLogger.debug("In onPause");

        if (!ignoreIncomingNotifications && MFPPush.getInstance() != null) {
            MFPPush.getInstance().hold();
        }
    }

}